### PR TITLE
Windows Minidump support

### PR DESCRIPTION
**Pull request by aarzilli**
```
proc/core: add support for windows minidumps

Minidumps are the windows equivalent of unix core files.
This commit updates pkg/proc/core so that it can open and read windows
minidumps.

Updates #794

proc: move windows register handling code to winutil package

Move windows register handling code to winutil package in preparation
for using it in proc/core to read minidump files.

```

### CODE REVIEW

**Comment by aarzilli** [comment]
I'm going to see if it's possible to convince appveyor to produce the minidump instead of having one in the repository.

**Comment by aarzilli** [comment]
Done.

File: pkg/proc/core/core.go [hunk 1]
```diff
@@ -184,11 +184,23 @@ var (
 	ErrChangeRegisterCore = errors.New("can not change register values of core process")
 )
 
+type openFn func(string, string) (*Process, error)
+
+var openFns = []openFn{readLinuxAMD64Core, readAMD64Minidump}
+var ErrUnrecognizedFormat = errors.New("unrecognized core format")
```
Comment (derekparker) [inline pkg/proc/core/core.go]: nit: Comment on this exported type.

File: _fixtures/sleep.go [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package main
```
Comment (derekparker) [inline _fixtures/sleep.go]: We have a fixture that sleeps indefinitely which we could use for this, we could just rename that file. It's `_fixtures/test1145.go`.

File: pkg/proc/core/core_test.go [hunk 3]
```diff
@@ -367,3 +367,88 @@ mainSearch:
...
+		t.Fatalf("could not find main goroutine")
+	}
+}
+
+func procdump(t *testing.T, exePath string) string {
+	exeDir := filepath.Dir(exePath)
+	cmd := exec.Command("procdump64", "-accepteula", "-ma", "-n", "1", "-s", "3", "-x", exeDir, exePath)
+	out, err := cmd.CombinedOutput() // procdump exits with non-zero status on success, so we have to ignore the error here
+	if !strings.Contains(string(out), "Dump count reached.") {
+		t.Fatalf("error running procdump64 (maybe), output: %q, error: %v", string(out), err)
```
Comment (derekparker) [inline pkg/proc/core/core_test.go]: nit: Reword as `possible error running procdump64, ...`

File: pkg/proc/core/minidump/minidump.go [hunk 4]
```diff
@@ -0,0 +1,689 @@
...
+	CpuArchitectureAMD64   Arch = 9
+	CpuArchitectureWoW64   Arch = 10
+	CpuArchitectureARM64   Arch = 12
+	CpuArchitectureUnknown Arch = 0xffff
+)
+
+// Open reads the minidump file at path and returns it as a Minidump structure.
+func Open(path string, logfn func(fmt string, args ...interface{})) (*Minidump, error) {
+	rawbuf, err := ioutil.ReadFile(path) //TODO(aarzilli): mmap?
+	must(err)
```
Comment (derekparker) [inline pkg/proc/core/minidump/minidump.go]: This convenience function is only used once, let's just inline the error check / panic.

File: pkg/proc/core/minidump/minidump.go [hunk 5]
```diff
@@ -0,0 +1,689 @@
...
+	end := off + int(sz)
+	if off >= len(buf.buf) || end > len(buf.buf) {
+		buf.err = fmt.Errorf("location starting at %#x of size %#x is past the end of file, while %s", off, sz, buf.ctx)
+		return 0, nil
+	}
+	rawData = buf.buf[off:end]
+	return
+}
+
+func readString(buf *minidumpBuf) string {
```
Comment (derekparker) [inline pkg/proc/core/minidump/minidump.go]: Could we just reuse `pkg/dwarf/util.ParseString` here?

File: pkg/proc/core/minidump/minidump.go [hunk 6]
```diff
@@ -0,0 +1,689 @@
...
+			return
+		}
+
+		nameBuf := minidumpBuf{buf: buf.buf, kind: "file", off: nameOff, err: nil, ctx: buf.ctx}
+		module.Name = readString(&nameBuf)
+		if nameBuf.err != nil {
+			buf.err = nameBuf.err
+			return
+		}
+
```
Comment (derekparker) [inline pkg/proc/core/minidump/minidump.go]: nit: whitespace cleanup

File: pkg/proc/core/minidump/minidump.go [hunk 7]
```diff
@@ -0,0 +1,689 @@
...
+	}
+}
+
+// readMiscInfo reads the process_id from a MiscInfo stream.
+func readMiscInfo(mdmp *Minidump, buf *minidumpBuf) {
+	buf.u32() // size of info
+	buf.u32() // flags1
+
+	mdmp.Pid = buf.u32() // process_id
+	// there are more fields here, but we don't care about them
```
Comment (derekparker) [inline pkg/proc/core/minidump/minidump.go]: Should we move the buffer past this memory, or are we simply done with the stream at this point?

File: pkg/proc/core/minidump/minidump.go (hunk 5 above)
Comment (aarzilli) [reply pkg/proc/core/minidump/minidump.go]: ParseString reads a NUL terminated string, this is a pascal-like string.

File: pkg/proc/core/minidump/minidump.go (hunk 7 above)
Comment (aarzilli) [reply pkg/proc/core/minidump/minidump.go]: We're done. The buffer used here is for the misc info stream only.

**Comment by aarzilli** [comment]
Done.

**Comment by derekparker** [comment]
@aarzilli there's still a few comments not addressed from the last review.

File: pkg/proc/core/minidump/minidump.go [hunk 8]
```diff
@@ -0,0 +1,689 @@
...
+		if logfn != nil {
+			logfn("Found processor architecture %s\n", arch.String())
+		}
+
+		if arch != CpuArchitectureAMD64 {
+			return nil, fmt.Errorf("unsupported architecture %s", arch.String())
+		}
+	}
+
+	for i := range mdmp.Streams {
```
[the remark this answers is not in the document]
Comment (aarzilli) [reply pkg/proc/core/minidump/minidump.go]: Some of the other streams have a different layout depending on the architecture, which could cause spurious error messages if we try to parse them first.

**Comment by aarzilli** [comment]
Sorry, I didn't see them because github helpfully collapsed them.

File: pkg/proc/core/minidump/minidump.go [hunk 9]
```diff
@@ -0,0 +1,688 @@
...
+// See: https://docs.microsoft.com/en-us/windows/desktop/api/minidumpapiset/ns-minidumpapiset-_minidump_memory64_list
+// And: https://docs.microsoft.com/en-us/windows/desktop/api/minidumpapiset/ns-minidumpapiset-_minidump_memory_descriptor
+func readMemory64List(mdmp *Minidump, buf *minidumpBuf, logfn func(fmt string, args ...interface{})) {
+	rangesNum := buf.u64()
+	baseOff := int(buf.u64())
+	if buf.err != nil {
+		return
+	}
+
+	_ = baseOff
```
Comment (derekparker) [inline pkg/proc/core/minidump/minidump.go]: Why this assignment?

Comment (aarzilli) [reply pkg/proc/core/minidump/minidump.go]: Leftover

**Pull request merged**